File: Packages/Learning/Sources/Learning/PersonalLexicon.swift
import Foundation
import Lexicon

/// Immutable `Lexicon`-conforming snapshot of a `PersonalModel`, so the
/// wave-2 engine can blend the personal store as a third suggestion source
/// through exactly the same protocol as the base `is.lex` / `en.lex`
/// readers (`FrequencyLexicon`) — no special-casing in the blender.
///
/// Why a wrapper instead of conforming `PersonalModel` directly:
/// `Lexicon` requires `Sendable`, and the model is a mutable reference type
/// that the app compacts on its own schedule. A value-type snapshot taken at
/// keyboard launch (and refreshed when the app signals a new model file) is
/// both trivially Sendable and immune to mid-keystroke mutation.
///
/// Only LEARNED words are exposed (threshold met, explicit signal, or
/// user-added — never tombstoned): pending one-off commits must not leak
/// into ranking. Bigram continuations are pair-level evidence and are
/// exposed as the model recorded them (minus tombstoned followers).
///
/// Personal scale is small (thousands of entries), so `completions` /
/// `continuations` are linear scans — measured in microseconds at this size,
/// mirroring the `DictLexicon` approach in TypeEngine.
public struct PersonalLexicon: Lexicon, Sendable {
    private let unigrams: [String: UInt32]
    private let bigrams: [String: UInt32]
    public let totalUnigramTokens: UInt64

    public init(model: PersonalModel) {
        var unigrams: [String: UInt32] = [:]
        for word in model.learnedWords {
            unigrams[word] = model.frequency(of: word) ?? 0
        }
        for word in model.userAddedWords {
            unigrams[word] = model.frequency(of: word) ?? 1
        }
        self.unigrams = unigrams

        var bigrams: [String: UInt32] = [:]
        for (key, count) in model.bigrams {
            guard let spaceIndex = key.firstIndex(of: " ") else { continue }
            let first = String(key[..<spaceIndex])
            let second = String(key[key.index(after: spaceIndex)...])
            guard !model.isTombstoned(first), !model.isTombstoned(second) else { continue }
            bigrams[key] = count
        }
        self.bigrams = bigrams

        totalUnigramTokens = unigrams.values.reduce(UInt64(0)) { $0 + UInt64($1) }
    }

    public func frequency(of word: String) -> UInt32? {
        unigrams[word]
    }

    public func bigramFrequency(_ first: String, _ second: String) -> UInt32? {
        bigrams["\(first) \(second)"]
    }

    public func completions(of prefix: String, limit: Int) -> [(word: String, frequency: UInt32)] {
        guard limit > 0 else { return [] }
        return unigrams
            .filter { $0.key.hasPrefix(prefix) }
            .sorted { $0.value > $1.value || ($0.value == $1.value && $0.key < $1.key) }
            .prefix(limit)
            .map { (word: $0.key, frequency: $0.value) }
    }

    public func continuations(of word: String, limit: Int) -> [(word: String, frequency: UInt32)] {
        guard limit > 0 else { return [] }
        let prefix = word + " "
        return bigrams
            .compactMap { key, freq -> (word: String, frequency: UInt32)? in
                guard key.hasPrefix(prefix) else { return nil }
                return (word: String(key.dropFirst(prefix.count)), frequency: freq)
            }
            .sorted { $0.frequency > $1.frequency || ($0.frequency == $1.frequency && $0.word < $1.word) }
            .prefix(limit)
            .map { $0 }
    }
}

File: AndroidClient/lib/engine/src/main/kotlin/is/solberg/lyklabord/engine/learning/PersonalLexicon.kt
package `is`.solberg.lyklabord.engine.learning

import `is`.solberg.lyklabord.engine.lexicon.LexEntry
import `is`.solberg.lyklabord.engine.lexicon.Lexicon

/** Immutable Lexicon snapshot of learned and user-added surface forms. */
class PersonalLexicon(model: PersonalModel) : Lexicon {
    private val unigrams: Map<String, UInt>
    private val bigrams: Map<String, UInt>
    override val totalUnigramTokens: ULong

    init {
        val uni = linkedMapOf<String, UInt>()
        model.learnedWords.forEach { uni[it] = model.frequency(of = it) ?: 0u }
        model.userAddedWords.forEach { uni[it] = model.frequency(of = it) ?: 1u }
        unigrams = uni
        val bi = linkedMapOf<String, UInt>()
        model.bigrams.forEach { (key, count) ->
            val space = key.indexOf(' ')
            if (space < 0) return@forEach
            val first = key.substring(0, space)
            val second = key.substring(space + 1)
            if (!model.isTombstoned(first) && !model.isTombstoned(second)) bi[key] = count
        }
        bigrams = bi
        totalUnigramTokens = unigrams.values.fold(0uL) { total, value -> total + value.toULong() }
    }

    override fun frequency(of: String): UInt? = unigrams[of]
    override fun bigramFrequency(first: String, second: String): UInt? = bigrams["$first $second"]
    override fun completions(of: String, limit: Int): List<LexEntry> =
        if (limit <= 0) emptyList() else unigrams.asSequence().filter { it.key.startsWith(of) }
            .sortedWith(compareByDescending<Map.Entry<String, UInt>> { it.value }.thenBy { it.key })
            .take(limit).map { LexEntry(it.key, it.value) }.toList()
    override fun continuations(of: String, limit: Int): List<LexEntry> =
        if (limit <= 0) emptyList() else bigrams.asSequence().filter { it.key.startsWith("$of ") }
            .map { LexEntry(it.key.removePrefix("$of "), it.value) }
            .sortedWith(compareByDescending<LexEntry> { it.frequency }.thenBy { it.word }).take(limit).toList()
}
